# CLASIFICACIÓN CON MÁQUINAS DE VECTORES DE SOPORTE - CÁNCER DE PECHO

En este laboratorio vamos a explorar el método de los vectores de soporte de clasificación. Para este objetivo vamos a emplear la base de datos de células potencialmente malignas que hemos utilizado en los laboratorios anteriores. 

In [1]:
# Importamos los paquetes que vayamos a usar

# Manejo de datos
import pandas as pd
import numpy as np

# Gráficos
import matplotlib as plt
import matplotlib.pyplot as plt
import seaborn as sns

# Preprocesado de datos y modelado
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC


In [2]:
# Cargamos la base de datos:

df = pd.read_csv('breast_cancer.csv')
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [3]:
# Generamos las variables con las que vamos a trabajar: 
X = df.iloc[:, 2:-1]
y = df['diagnosis']

# Separamos los datos en subgrupos de entrenamiento y testeo
x_train,x_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=0)
# Random_state nos creara una semilla aleatoria para tener el mismo resultado cada vez que corramos el codigo


In [4]:
# Definimos el algoritmoo con el que vamos a trabajar: 

algoritmo = SVC()

In [5]:
# Entrenamos el modelo:

algoritmo.fit(x_train, y_train)

SVC()

In [6]:
# Predecimos con el modelo entrenado:

predichos = algoritmo.predict(x_test)
predichos

array(['B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B',
       'M', 'B', 'M', 'B', 'M', 'M', 'M', 'M', 'M', 'B', 'B', 'M', 'B',
       'B', 'B', 'B', 'M', 'B', 'M', 'B', 'M', 'B', 'B', 'B', 'M', 'B',
       'M', 'B', 'B', 'M', 'B', 'B', 'M', 'B', 'B', 'B', 'M', 'M', 'B',
       'M', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'M', 'B', 'B', 'M',
       'B', 'M', 'M', 'M', 'B', 'B', 'M', 'B', 'B', 'M', 'B', 'B', 'B',
       'B', 'B', 'M', 'M', 'M', 'B', 'M', 'B', 'B', 'B', 'M', 'M', 'B',
       'M', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B',
       'M', 'B', 'M', 'B', 'M', 'M', 'B', 'M', 'M', 'B', 'B', 'B', 'B',
       'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B',
       'M', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'M', 'B', 'B', 'B', 'M',
       'B', 'B', 'M', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B',
       'B', 'B', 'B', 'M', 'M', 'B', 'B', 'M', 'B', 'M', 'B', 'M', 'B',
       'B', 'B'], dtype=object)

In [7]:
# Evaluamos como de bueno es el modelo

# Matriz confunsión:
pd.crosstab(y_test, predichos)


col_0        B   M
diagnosis         
B          107   1
M           12  51

In [8]:
# Creamos un diccionario vacio para nuestros resultados
Evaluaciones = {}

prediccion = np.array(predichos)

# EXACTITUD
# Calculamos la exactitud del modelo
from sklearn.metrics import accuracy_score
exactitud =accuracy_score(y_test, prediccion)

# Lo incorporamos a nuestro diccionario
Evaluaciones["exactitud"] = exactitud

# PRECISIÓN
#Calculamos la Precisión del modelo
from sklearn.metrics import precision_score
precision =precision_score(y_test, prediccion, pos_label='M')

# Lo incorporamos a nuestro diccionario
Evaluaciones["precision"] = precision

# SENSIBILIDAD
# Calculamos la sensibilidad del modelo
from sklearn.metrics import recall_score
sensibilidad =recall_score(y_test, predichos, pos_label='M')

# Lo incorporamos a nuestro diccionario
Evaluaciones["sensibilidad"] = exactitud


# EXACTITUD
# Calculamos la exactitud del modelo
from sklearn.metrics import f1_score
F1 =f1_score(y_test, predichos, pos_label='M')

# Lo incorporamos a nuestro diccionario
Evaluaciones["F1"] = F1

In [9]:
Evaluaciones

{'exactitud': 0.9239766081871345,
 'precision': 0.9807692307692307,
 'sensibilidad': 0.9239766081871345,
 'F1': 0.8869565217391304}

Como deciamos en los laboratorios anteriores, la mejor métrica para evaluar este analisis puede que sea la sensibilidad. Ya que nos da constancia de los positivos identificados correctamente entre los positivos reales. Es decir, la cantidad células malignas que identificamos correctamente como malignas. En este caso, la sensibilidad es del 92,4%, lo que significaría que dariamos por benigna aproximadamente 1 de cada 10 celulas malignas. Teniendo en cuenta el contexto, considero que no sería un modelo deseable. De hecho, el modelo logaritmo sigue dando el mejor resultado con una sensibilidad del 98%.